File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;

        stack<pair<TreeNode*, int>> stack;
        stack.push({root, 1});

        int max = 0;
        while (!stack.empty()) {
            pair<TreeNode*, int> current = stack.top();
            stack.pop();

            TreeNode* node = current.first;
            int depth = current.second;

            if (node != nullptr) {
                max = std::max(depth, max);
                stack.push({node -> right, depth + 1});
                stack.push({node -> left, depth + 1});
            }

        }

        return max;
        
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# longest length between any two roots
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
         
        m = 0
        def dfs(r):
            nonlocal m
            if not r:
                return 0

            left = dfs(r.left)
            right = dfs(r.right)

            m = max(m, left + right)

            return 1 + max(left, right)


        dfs(root)
        return m
    